File: lib/php/GMRRequest.php
<?php
final class GMRRequest
{
	private $options;
	private $key;
	private $masquerade;
	const kDomain = 'http://hazgame.com:7331';
	
	
	public function __construct($key, $masquerade=null)
	{
		$this->key        = $key;
		$this->masquerade = $masquerade;
		
	}
	
	public function execute($options)
	{
		$headers = array('Authorization: OAuth '.$this->key,
		                 'Accept: application/json');
		
		
		if($this->masquerade)
			array_push($headers, "X-Masquerade-As: ".$this->masquerade);
		
		$request = curl_init();
		
		$url = self::kDomain.$options['path'];
		
		if(isset($options['query']))
			$url .= '?'.http_build_query($options['query']);
		
		curl_setopt($request, CURLOPT_URL, $url);
		curl_setopt($request, CURLOPT_HEADER, false);
		curl_setopt($request, CURLOPT_RETURNTRANSFER, true);
		curl_setopt($request, CURLOPT_HTTPHEADER, $headers);
		
		switch($options['method'])
		{
			case 'POST':
				curl_setopt($request, CURLOPT_POST, true);
				$data = null;
				
				if(isset($options['data']))
					$data = is_array($options['data']) ? http_build_query($options['data']) : $options['data'];
				
				curl_setopt($request, CURLOPT_POSTFIELDS, $data);
				break;
			
			case 'DELETE':
				curl_setopt($request, CURLOPT_CUSTOMREQUEST, 'DELETE');
				break;
		}
		
		$response = curl_exec($request);
		curl_close($request);
		return $response;
	}
	
}
?>

File: lib/ios/Gamer/GMRRequest.m
//
//  GMRRequest.m
//  Gamer
//
//  Created by Adam Venturella on 11/2/10.
//  Copyright 2010 __MyCompanyName__. All rights reserved.
//

#import "GMRRequest.h"
#import "GMRTypes.h"
#import "GMRUtils.h"
#import "ASIHTTPRequest.h"
#import "YAJLDocument.h"
#include <dispatch/dispatch.h>


#define API_DOMAIN @"http://hazgame.com:7331"
#define USER_AGENT @"HazGame Mobile"

#define GAMER_TESTING 1

static dispatch_queue_t networkQueue;
static dispatch_queue_t jsonProcessingQueue;

@implementation GMRRequest
@synthesize key;

- (id)init
{
	self = [super init];
	if(self)
	{
		networkQueue        = dispatch_queue_create("com.hazgame.network", NULL);
		jsonProcessingQueue = dispatch_queue_create("com.hazgame.json", NULL);
	}
	
	return self;
}

- (void)execute:(NSDictionary *)options withCallback:(GMRCallback)callback
{
	//NSOperationQueue * q = [[[NSOperationQueue alloc] init] autorelease];
	
	dispatch_async(networkQueue, ^{
		
		NSString * path   = (NSString *)[options objectForKey:@"path"];
		NSString * method = (NSString *)[options objectForKey:@"method"];
		
		NSRange absolutePath = [path rangeOfString:@"http://"];
		
		if(absolutePath.location == NSNotFound)
			path = [NSString stringWithFormat:@"%@%@", API_DOMAIN, path];
			
		
		NSString * endpoint = path;
		
		if ([options objectForKey:@"query"] != nil) 
		{
			NSDictionary * query = (NSDictionary *) [options objectForKey:@"query"];
			endpoint = [NSString stringWithFormat:@"%@?%@", endpoint, [GMRUtils formEncodedStringFromDictionary:query]];
		}
		
		
		NSURL * url = [NSURL URLWithString:endpoint];
		ASIHTTPRequest *request = [ASIHTTPRequest requestWithURL:url];
		
		[request addRequestHeader:@"Authorization" value:[NSString stringWithFormat:@"OAuth %@", self.key]];
		[request addRequestHeader:@"Accept"        value:@"application/json"];
		[request addRequestHeader:@"User-Agent"    value:USER_AGENT];
		
		[request setRequestMethod:method];
		
		if([method isEqualToString:@"POST"])
		{
			NSDictionary * data = [options objectForKey:@"data"];
			
			if(data)
			{
				NSMutableData * postBody = [NSMutableData data];
				
				[postBody appendData:[[GMRUtils formEncodedStringFromDictionary:data] dataUsingEncoding:NSUTF8StringEncoding]];
				
				[request addRequestHeader:@"Content-Type"    value:@"application/x-www-form-urlencoded"];
				[request addRequestHeader:@"Content-Length"  value:[NSString stringWithFormat:@"%u", [postBody length]]];
				[request setPostBody:postBody];
			}
			
		}
		
		[ASIHTTPRequest setShouldUpdateNetworkActivityIndicator:NO];
		[request startSynchronous];
		
		
		dispatch_async(jsonProcessingQueue, ^{
			NSError * error = nil;
			YAJLDocument * json = [[[YAJLDocument alloc] initWithData:[request responseData] 
														parserOptions:YAJLParserOptionsNone 
																error:&error] autorelease];		
			if(error)
			{
				NSLog(@"GMRRequest Network Error!");
				NSLog(@"%@",[request responseString]);
			}
				
			
			NSLog(@"%@", (NSDictionary *)json.root);
			callback([[json.root objectForKey:@"ok"] boolValue], (NSDictionary *)json.root);
		});
		
	});
}


- (void)dealloc
{
	dispatch_release(networkQueue);
	dispatch_release(jsonProcessingQueue);
	self.key = nil;
	
	[super dealloc];
}
@end
